Merge for "Histogram on actors with map-reduce approach" task

nikodrozd opened the pull request.

nikodrozd committed:
Added solution and test for task "Histogram on actors with map-reduce approach"

tanzan commented on src/main/scala/histogramactors/HistogramMain.scala:
```diff
@@ -0,0 +1,137 @@
+  override def receive: Receive = {
+    case Load(filePath) =>
+      loadData(filePath) match {
+        case Success(dataFromFile) => sender() ! LoadResult(dataFromFile)
```
Bad practice calling blocking operations in receive, try using a future.

nikodrozd replied:
Call of loadData function has been added to Future

tanzan commented on src/main/scala/histogramactors/HistogramMain.scala:
```diff
@@ -0,0 +1,137 @@
+class Master(input: String, output: String, system: ActorSystem) extends Actor {
+  private val poolSize = 10
+
+  val loader: ActorRef = system.actorOf(Props[Loader](), "loader")
```
Common practice creating child actors using `context.actoryOf`

nikodrozd replied:
Changed to proposed variant

tanzan commented on src/main/scala/histogramactors/HistogramMain.scala:
```diff
@@ -0,0 +1,137 @@
+
+}
+
+class Master(input: String, output: String, system: ActorSystem) extends Actor {
```
Try simplifying design:

-  there are too many very fine-grained actors
- overall design is computationally ineffective (I guess messaging overhead is greater than the job itself), try splitting a file into batches, process each batch using the mapper, then combine results at the reducer.

nikodrozd replied:
Reworked in following way:
1) sequence of lines are grouped to blocks with size = (lines.size / 10) + 1
2) Line case class changed to Block with Seq[String] as input
3) Parser works with whole block of lines now and sends result to reducer

nikodrozd committed:
Code has been refactored after review:
1) changed creation of child actors
2) added Future to call of loadData function
3) mapper has been changed to work with blocks of text instead of work with each line

tanzan commented on src/main/scala/histogramactors/HistogramMain.scala:
```diff
@@ -0,0 +1,137 @@
+  val outputFilePath = "src/main/resources/histogram/output.txt"
+
+  val system = ActorSystem("histogram")
+  val master = system.actorOf(Props(classOf[Master], inputFilePath, outputFilePath, system), "master")
```
You don't need to pass the system to an actor. You could always get it from the actor context.

tanzan commented on src/main/scala/histogramactors/HistogramMain.scala:
```diff
@@ -49,9 +49,17 @@ class Master(input: String, output: String, system: ActorSystem) extends Actor {
+          case Success(dataFromFile) => LoadResult(dataFromFile)
+          case Failure(exception) => FailResult(exception)
+        }
+      } onComplete {
```
You can use pipeTo

tanzan commented on src/main/scala/histogramactors/HistogramMain.scala:
```diff
@@ -24,11 +24,11 @@ object HistogramMain extends App {
+  val saver: ActorRef = context.actorOf(Props[Saver](), "saver")
   override def receive: Receive = {
     case Start =>
       loader ! Load(input)
```
It is possible to start a few jobs. You need to prevent this either redesign it in a way that it's possible to run a few files simultaneously. In another way that would be a complete mess.

tanzan commented on src/main/scala/histogramactors/HistogramMain.scala:
```diff
@@ -0,0 +1,137 @@
+    case FailResult(ex) =>
+      throw ex
+      system.terminate()
+    case SuccessResult =>
```
It's not a good idea to terminate the actor system inside the actor. It's like doing `System.exit` somewhere in code.

tanzan replied:
Try to figure out a solution when a single failure doesn't lead to the complete stop of a job but if some task is failed then it could be restarted. Please utilize `SupervisionStrategy`. In the end, after few restarts could terminate the master, not the actor system. You can 'master.ask(Start).onComplete(...   print errors, system.terminate  ...)'

nikodrozd committed:
Code has been refactored after review:
1) removed Loader, Saver, Mapper and ParserPool actors
2) added retry functionality in case of Parser fail
